### PR TITLE
Feature: Export Processed Requests as CSV

### PR DESCRIPTION
**Pull request by yongkanm**
### Description

This pull request implements the feature of exporting processed requests as CSV. This feature has three entry points:

1.  The “Export Processed Requests” button at the top-right corner of the “Token Options” page. This entry point will export processed requests for all token options.
2.  The “Export Processed Requests” button in the drop-down menu of a token option group. This entry point will only export processed requests for this token option group.
3.  The “Export Processed Requests” button in the drop-down menu of a token option. This entry point will only export processed requests for this token option.

When exporting requests, the user could choose the scope of the processed requests. Currently, three types of scopes are supported:

1.  Entire Course (in one CSV): A single CSV file will be generated.
2.  Entire Course (one CSV per section): A zip file that contains one or more CSVs will be generated. Each CSV file corresponds to a section. A CSV file will be created only if students from the corresponding section have processed request(s) that satisfy the specified restrictions.
3.  One section only: The user could choose a section. A single CSV file, which only contains processed requests that both satisfy the specified restrictions and are made by students from the selected section, will be generated.

In addition, the user could configure a time range: only processed requests made within this time range will be exported.

If Token ATM found that there are no requests satisfy the specified restrictions, no file will be generated. Instead, a message will be shown to indicate this fact.

If errors occurred during the exporting process, the exporting process will be terminated, while a modal including the error message will be shown.

### Testing Note

Please test if the feature works as described above.

### CODE REVIEW

**Comment by yongkanm** [comment]
> **Note**: Users can change export scope, but the save file button for the previously exported scope is still visible. This could break user expectation, because a save file button exists but not for the selected scope.
> 
> Possible fixes:
> 
> 1. Track a generated file blob for each export scope and display the save file button if the mapping for that scope and file pair exists. (Requires another data structure and utilities to prioritize and clear some/all file refs. Extra care should be taken about memory usage. Maybe only allow a single scope->file pair at a time, but hide the save button if that specific scope isn't selected.)
> 2. Automatically clear file after export scope change. (Reprocessing will be required after each scope selection, and silently removing generated files may confuse/frustrate users.)
> 3. Buttons/Interactions that lock/unlock scope selection change and/or warns of file being removed after. (Interrupts a simple/smooth interaction flow)

Thanks for finding out this possible use case. I decide to go with option 3: add a new stage between the completion of the first export and the start of the second export. The user now has to first click a button to clear the current export and then modify the fields.

File: src/app/components/export-request-modal/export-request-modal.component.ts [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+                                    return false;
+                                }
+                                return true;
+                            };
+                        })
+                )
+                .appendBuilder(
+                    createFieldComponentWithLabel(
+                        CheckboxFieldComponent,
+                        'Include Rejected Request',
```
[the remark this answers is not in the document]
Comment (yongkanm) [reply src/app/components/export-request-modal/export-request-modal.component.ts:140]: Thanks! Change is included in the next commit.

**Comment by yongkanm** [comment]
> **Note**: The exported filenames are helpful and with the timestamp are unique. But more informative names would be useful. I.e., including entry point scope, export type scope, if rejected requests are included, or other configured/selected information.
> 
> It may be worthwhile to also think about creating a zip for each download that contains a text file with the configurations the export used as meta info. That would prevent the ballooning of filename length while also providing details about the export.
> 
> Maybe the root zip includes a format of something like `{prefix}-{course name}-{time stamp}` and all subfiles/subfolders of the zip use the same time stamp to help resolve/identify that a group of individual unzipped files are all part of the same export.

I've thought about metadata file, but it complicates the information access process. I'm still not sure about how should we optimize this. Timestamp is handy, but it's definitely not user-friendly. I think we could gather some user feedback first and enhance it later.

File: src/app/services/request-exporter.service.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+                    continue;
+                }
+                const groupName = await this.classifier(request, student);
+                if (!requestDataMap.has(groupName)) requestDataMap.set(groupName, []);
+                requestDataMap.get(groupName)?.push(parseProcessedRequest(request));
+            }
+            this.curProgress++;
+        }
+        const fileName = `Token-ATM-Request-Export-${
+            this.fileName ? sanitizeFileName(this.fileName) + '-' : '' + format(new Date(), 'MM-dd-yyyy-HH-mm-ss')
```
[the remark this answers is not in the document]
Comment (yongkanm) [reply src/app/services/request-exporter.service.ts:80]: Thanks! Change is included in the next commit. For timestamp, please check out my comment regarding exported filename.

**Review by epsilonError — DISMISSED** [review]

**Review by epsilonError — APPROVED** [review]

**Pull request merged**